exchanges: Add V2 Crypto.com exchange support

samuael opened the pull request:
# PR Description

This pull request adds a code implementation for version 2 of the Crypto.com exchange. The implementation supports REST and web-socket functionalities. All the REST, WebSocket endpoints, and wrapper functions are tested. My code adheres to the GCT style standards used by other exchanges. 
Fixed some errors with the help of linter(golangci-lint), and unit tests. REST and WebSocket endpoint methods that have no support from the GCT wrapper are also implemented for future use.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as the item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also, consider improving test coverage whilst working on a certain feature or package.

- [x] go test ./... -race
- [x] golangci-lint run
- [x] Test X

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,966 @@
+	if len(tick.Data) != 1 {
+		return tickerPrice, errInvalidResponseFromServer
+	}
+	tickerPrice = &ticker.Price{
```
`new(ticker.Price)` allocates and is a non-nil value. We can just remove that above and do a short hand `:=` operator.

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,966 @@
+			Pair:         cp,
+			ExchangeName: cr.Name,
+			AssetType:    assetType,
+			// QuoteVolume:  tick.Data[y].QuoteVolume,
```
Uncomment?

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,966 @@
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	var notional float64
+	switch s.Type {
```
NOTE: Test for when both s.Amount and notional are sent.

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,966 @@
+
+// GetHistoricCandles returns candles between a time period for a set time interval
+func (cr *Cryptodotcom) GetHistoricCandles(ctx context.Context, pair currency.Pair, a asset.Item, interval kline.Interval, _, _ time.Time) (*kline.Item, error) {
+	if pair.IsEmpty() {
```
Please use GetKlineRequest and utilize the start and end times, if it exceeds 200 candles it will error.

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,966 @@
+
+// GetHistoricCandlesExtended returns candles between a time period for a set time interval
+func (cr *Cryptodotcom) GetHistoricCandlesExtended(ctx context.Context, pair currency.Pair, a asset.Item, interval kline.Interval, _, _ time.Time) (*kline.Item, error) {
+	return cr.GetHistoricCandles(ctx, pair, a, interval, time.Time{}, time.Time{})
```
if date ranges are not used we can errFunctionNotSupport.

shazbert commented on testdata/configtest.json:
```diff
@@ -2322,7 +2591,10 @@
+     "websocketAPI": false,
+     "saveTradeData": false,
+     "tradeFeed": false,
+     "fillsFeed": false
```
probably revert all these changes.

shazbert reviewed (changes requested):
Looks good! Just a basic look over. Well done.

samuael commented on exchanges/cryptodotcom/cryptodotcom.go:
```diff
@@ -0,0 +1,914 @@
+// GetCurrencyNetworks retrives the symbol network mapping.
+func (cr *Cryptodotcom) GetCurrencyNetworks(ctx context.Context) (*CurrencyNetworkResponse, error) {
+	var resp *CurrencyNetworkResponse
+	return resp, cr.SendAuthHTTPRequest(ctx, exchange.RestSpot, privateGetCurrencyNetworksRate, privateGetCurrencyNetworks, nil, &resp)
```
We can not unmarshal nil a binary into the pointer interface. I have to use the pointer of the filed.

shazbert replied:
That's fair I will resolve.

shazbert commented on exchanges/cryptodotcom/cryptodotcom.go:
```diff
@@ -740,7 +658,9 @@ func (cr *Cryptodotcom) SendAuthHTTPRequest(ctx context.Context, ePath exchange.
+	response := &RespData{
+		Result: resp,
+	}
 	newRequest := func() (*request.Item, error) {
```
[Previous nit](https://github.com/thrasher-corp/gocryptotrader/pull/1122#discussion_r1090143016)

Incomplete.

shazbert commented on exchanges/cryptodotcom/cryptodotcom.go:
```diff
@@ -835,7 +752,7 @@ func (cr *Cryptodotcom) getParamString(params map[string]interface{}) string {
 			param, ok := params[keys[x]].(map[string]interface{})
 			if !ok {
-				param = map[string]interface{}{}
+				param = make(map[string]interface{})
```
In another PR I have suggested we use the switch case to access an interfaces underlying inbuilt type. We can use that here instead of using reflection which can be slow in some instances.

shazbert commented on exchanges/cryptodotcom/cryptodotcom_test.go:
```diff
@@ -252,6 +252,7 @@ func TestGetPersonalOpenOrders(t *testing.T) {
 	if err != nil {
 		t.Error(err)
 	}
+	t.Error(err)
```
remove

shazbert commented on exchanges/cryptodotcom/cryptodotcom_test.go:
```diff
@@ -413,7 +414,7 @@ func TestGetOTCQuoteHistory(t *testing.T) {
 		t.Skip(credInfoNotProvided)
 	}
-	_, err := cr.GetOTCQuoteHistory(context.Background(), currency.EMPTYCODE, currency.EMPTYCODE, time.Time{}, time.Time{}, 0, 10)
+	_, err := cr.GetOTCQuoteHistory(context.Background(), currency.NewPair(currency.EMPTYCODE, currency.EMPTYCODE), time.Time{}, time.Time{}, 0, 10)
```
currency.NewPair(currency.EMPTYCODE, currency.EMPTYCODE) -> currency.EmptyPair

shazbert commented on exchanges/cryptodotcom/cryptodotcom_types.go:
```diff
@@ -320,6 +300,85 @@ type CreateOrderParam struct {
 	TriggerPrice   float64    `json:"trigger_price"`
 }
 
+func (arg *CreateOrderParam) getCreateParamMap() (map[string]interface{}, error) {
```
This needs full test coverage and utilizing `if !errors.Is(err, errGlobalError) { }`

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -953,7 +946,7 @@ func (cr *Cryptodotcom) GetHistoricCandles(ctx context.Context, pair currency.Pa
 	}
 	return &kline.Item{
```
use kline.Request method

shazbert reviewed (changes requested):
Thanks for that, just a quick review of last changes. Once completed I will start doing end point tests.

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,953 @@
+	}
+	pairs := make(currency.Pairs, len(instruments))
+	for x := range instruments {
+		cp, err := currency.NewPairFromString(instruments[x].InstrumentName)
```
assign straight away 
```go
pairs[x], err = currency.NewPairFromString(instruments[x].InstrumentName)
```

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,953 @@
+		Asset:           assetType,
+		VerifyOrderbook: cr.CanVerifyOrderbook,
+	}
+	book.Bids = make([]orderbook.Item, len(orderbookNew.Data[0].Bids))
```
len check needed for `orderbookNew.Data` as this can panic if no data returned.

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,953 @@
+}
+
+// GetRecentTrades returns the most recent trades for a currency and asset
+func (cr *Cryptodotcom) GetRecentTrades(ctx context.Context, p currency.Pair, assetType asset.Item) ([]trade.Data, error) {
```
The assetype is not being checked, please run wrapper checks in the cmd folder to see any issues. There are wrapper functions that are not handling formatting pairs.

shazbert commented on exchanges/cryptodotcom/cryptodotcom_types.go:
```diff
@@ -0,0 +1,703 @@
+
+// SubscriptionRawData represents a subscription response raw data.
+type SubscriptionRawData struct {
+	Data          stream.Response
```
Just use `[]byte` we don't need to keep allocating `stream.Response` every data event

shazbert commented on exchanges/cryptodotcom/cryptodotcom_websocket.go:
```diff
@@ -0,0 +1,626 @@
+var responseStream chan SubscriptionRawData
+
+// WsConnect creates a new websocket to public and private endpoints.
+func (cr *Cryptodotcom) WsConnect() error {
```
Websocket isn't working.

shazbert reviewed (changes requested):
Thanks for changing the requested will go over authenticated endpoints once websocket is up and running.

shazbert commented on exchanges/cryptodotcom/cryptodotcom_test.go:
```diff
@@ -778,30 +780,30 @@ func TestWsSetCancelOnDisconnect(t *testing.T) {
 func TestGetCreateParamMap(t *testing.T) {
 	t.Parallel()
-	arg := &CreateOrderParam{InstrumentName: "BTC_USDT", ClientOrderID: "", TimeInForce: "", Side: order.Buy, OrderType: order.Limit, PostOnly: false, TriggerPrice: 0, Price: 123, Quantity: 12, Notional: 0}
+	arg := &CreateOrderParam{InstrumentName: "BTC_USDT", Side: order.Buy, OrderType: orderTypeToString(order.Limit), Price: 123, Quantity: 12}
```
[Previous nit not complete](https://github.com/thrasher-corp/gocryptotrader/pull/1122#discussion_r1099327224)

shazbert commented on exchanges/cryptodotcom/cryptodotcom_test.go:
```diff
@@ -491,7 +493,8 @@ func TestUpdateOrderbook(t *testing.T) {
 		t.Fatal(err)
 	}
-	_, err = cr.UpdateOrderbook(context.Background(), enabledPairs[0], asset.Spot)
+	cr.Verbose = true
```
remove

shazbert commented on exchanges/cryptodotcom/cryptodotcom_test.go:
```diff
@@ -479,6 +480,7 @@ func TestFetchOrderbook(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
+	cr.Verbose = true
```
remove

shazbert commented on exchanges/cryptodotcom/cryptodotcom_types.go:
```diff
@@ -312,18 +315,22 @@ func (arg *CreateOrderParam) getCreateParamMap() (map[string]interface{}, error)
+	}
+	if oType == order.UnknownType || oType == order.AnyType {
 		return nil, fmt.Errorf("%w, Order Type: %v", order.ErrTypeIsInvalid, arg.OrderType)
 	}
```
This can be shifted to the stringToOrderType function and error return.

shazbert commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -605,17 +672,32 @@ func (cr *Cryptodotcom) SubmitOrder(ctx context.Context, s *order.Submit) (*orde
-	})
+	var ordersResp *CreateOrderResponse
+	if cr.Websocket.CanUseAuthenticatedWebsocketForWrapper() {
+		ordersResp, err = cr.WsPlaceOrder(&CreateOrderParam{
+			InstrumentName: format.Format(s.Pair),
+			Side:           s.Side,
+			OrderType:      orderTypeToString(s.Type),
+			Price:          s.Price,
+			Quantity:       s.Amount,
+			ClientOrderID:  s.ClientOrderID,
+			Notional:       notional,
+			PostOnly:       s.PostOnly,
+			TriggerPrice:   s.TriggerPrice,
+		})
```
create a variable for CreateOrderParam pointer above so it doesn't need to be declared twice.

shazbert reviewed (changes requested):
Just some pending things on recent changes before I do function testing.

gloriousCode commented on exchanges/cryptodotcom/cryptodotcom_websocket.go:
```diff
@@ -0,0 +1,604 @@
+
+// GenerateDefaultSubscriptions Adds default subscriptions to websocket to be handled by ManageSubscriptions()
+func (cr *Cryptodotcom) GenerateDefaultSubscriptions() ([]stream.ChannelSubscription, error) {
+	subscriptions := []stream.ChannelSubscription{}
```
```suggestion
	var subscriptions []stream.ChannelSubscription
```
The reason is because if there are any errors, we're not allocating memory space. Just declaring it as a var is an empty declaration. Its minor, but its a good practice to continue doing in all PRs

gloriousCode commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,1049 @@
+
+// UpdateTicker updates and returns the ticker for a currency pair
+func (cr *Cryptodotcom) UpdateTicker(ctx context.Context, p currency.Pair, assetType asset.Item) (*ticker.Price, error) {
+	if !cr.SupportsAsset(assetType) {
+		return nil, fmt.Errorf("asset type of %s is not supported by %s", assetType, cr.Name)
+	}
```
```suggestion
if !cr.SupportsAsset(assetType) {
		return fmt.Errorf("%w asset type: %v", asset.ErrNotSupported, assetType)
	}
```

gloriousCode commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -0,0 +1,1049 @@
+
+// GetFeeByType returns an estimate of fee based on the type of transaction
+func (cr *Cryptodotcom) GetFeeByType(ctx context.Context, feeBuilder *exchange.FeeBuilder) (float64, error) {
+	return 0, common.ErrNotYetImplemented
```
Please implement

gloriousCode commented on exchanges/cryptodotcom/cryptodotcom_websocket_eps.go:
```diff
@@ -0,0 +1,282 @@
+package cryptodotcom
```
What does eps mean?

samuael replied:
endpoints!

gloriousCode commented on exchanges/cryptodotcom/cryptodotcom.go:
```diff
@@ -0,0 +1,816 @@
+		params.Set("depth", strconv.FormatInt(depth, 10))
+	}
+	var resp *OrderbookDetail
+	return resp, cr.SendHTTPRequest(ctx, exchange.RestSpot, common.EncodeURLValues(publicOrderbook, params), request.Unset, &resp)
```
You are not setting the ratelimits. I understand that this exchange has different style of ratelimits, but it seems you have defined them all but don't use them

gloriousCode commented on exchanges/cryptodotcom/cryptodotcom_websocket.go:
```diff
@@ -0,0 +1,604 @@
+// websocket subscriptions channels list
+
+const (
+	// private subscription channels
+	userOrderCnl   = "user.order.%s" // user.order.{instrument_name}
+	userTradeCnl   = "user.trade.%s" // user.trade.{instrument_name}
+	userBalanceCnl = "user.balance"
+
+	// public subscription channels
+	instrumentOrderbookCnl = "book.%s"           // book.{instrument_name}
+	tickerCnl              = "ticker.%s"         // ticker.{instrument_name}
+	tradeCnl               = "trade.%s"          // trade.{instrument_name}
+	candlestickCnl         = "candlestick.%s.%s" // candlestick.{time_frame}.{instrument_name}
```
Just use string joining since there isn't anything complicated here. eg `tickerCnl + "." + currPair`

gloriousCode reviewed (changes requested):
This looks pretty good! I'm in the process of getting API access, so just a quick overall code quality check first

shazbert reviewed (changes requested):
Thanks for making the changes, this has a data race in websocket.

samuael commented on exchanges/cryptodotcom/cryptodotcom_wrapper.go:
```diff
@@ -267,7 +267,11 @@ func (cr *Cryptodotcom) UpdateTradablePairs(ctx context.Context, forceUpdate boo
-		return nil, fmt.Errorf("asset type of %s is not supported by %s", assetType, cr.Name)
+		return nil, fmt.Errorf("%w asset type: %v", asset.ErrNotSupported, assetType)
+	}
+	p, err := cr.FormatExchangeCurrency(p, asset.Spot)
```
No, since crypto.com only use Spot, there maybe some places where I used the terms interchangeably. But, I will fix that

shazbert reviewed (changes requested):
There are a lot of error messages that can be seen when running the application:

```
[ERROR] | WEBSOCKET | 17/03/2023 15:00:06 | exchange Cryptodotcom websocket error - could not match incoming message with signature: 146884893, and data: {"id":146884893,"code":0,"method":"subscribe","channel":"book.RSR_USDT"}
```

gloriousCode commented on exchanges/cryptodotcom/cryptodotcom_websocket.go:
```diff
@@ -0,0 +1,613 @@
+)
+
+var defaultSubscriptions = []string{
+	// instrumentOrderbookCnl,
+	// tickerCnl,
```
Uncomment

gloriousCode commented on exchanges/cryptodotcom/cryptodotcom_websocket.go:
```diff
@@ -0,0 +1,613 @@
+	if err != nil {
+		return err
+	}
+	tickersDatas := make([]ticker.Price, len(data))
```
This doesn't do anything because `websocketroutine_manager.go` has no way of handling `[]ticker.Price` so the data goes nowhere, forcing a REST switchover.

You'll need to update `websocketroutine_manager.go` to support it in the function `websocketDataHandler`

shazbert commented on exchanges/cryptodotcom/cryptodotcom_websocket.go:
```diff
@@ -0,0 +1,613 @@
+		if err != nil {
+			return err
+		}
+		time.Sleep(time.Second)
```
You pause everything up >= 20 seconds. Rate limiting is done within SendJSONMesage function. This was hard to find, running the application would have shown there was an issue.

shazbert reviewed (changes requested):
Thanks some issues I have found. I will be opening up a PR to circumvent an edge case with respect to the GetHistoricCandles function shortly.

samuael commented:
Thanks for your PR. And, I was also wondering whether to have such a variable to distinguish whether asset Items( instruments of asset items) have candlestick data necessarily. Since some instruments may not have candlestick data even with in an extended time range

shazbert commented on exchanges/cryptodotcom/cryptodotcom_test.go:
```diff
@@ -0,0 +1,929 @@
+	var resp *QuoteHistoryResponse
+	err := json.Unmarshal([]byte(getOTCQuoteHistoryResponseJSON), &resp)
+	if err != nil {
+		t.Fatal(err)
```
--- FAIL: TestGetOTCQuoteHistory (0.00s)
    cryptodotcom_test.go:425: timestamp information of type <nil> is not supported
FAIL

shazbert reviewed (changes requested):
One issue before auth testing

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Added TimeInForce type and updated related files

samuael committed:
Linter issue fix and minor coinbasepro type update

samuael committed:
Bitrex consts update

samuael committed:
added unit test and minor changes in bittrex

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
add missing wrapper declaration

samuael committed:
Unit tests update

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Fix minor linter issues

samuael committed:
Update TestStringToTimeInForce unit test

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Types, Wrapper functions, and unit tests updates

samuael committed:
Minor unit test update

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Types update

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptocom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
fix conflict with gateio timeInForce

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Update unit test handlings

samuael committed:
Fit unix test, types, and linter issues

samuael committed:
Unit tests and few endpoints update

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Update ratelimiter, wrapper, and few changes

samuael committed:
Config lint fix

samuael committed:
endpoints update and started adding missing endpoints, and unit tests

samuael committed:
Added missing endpoints and unit tests

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Minor cleanup on websocket configuration

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
replace errors and minor changes

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Update REST endpoints, types, and unit tests

samuael committed:
Websocket endpoints error handling update

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Added Staking endpoint and unit testing

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Added missing updates

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Update websocket endpoints, types, and unit tests

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
wrapper setup and endpoint fix

samuael committed:
update endpoints and unit tests

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Update order tests

samuael committed:
Complete updating the order unit tests

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif # Please enter a commit message to
explain why this merge is necessary, # especially if it merges an updated upstream into a topic branch. # # Lines starting
with '#' will be ignored, and an empty message aborts # the commit.

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
minor fix on urls and unit test

samuael committed:
Merge branch 'cryptoCom' of https://github.com/samuael/gocryptotrader; branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
update kucoin and deribit wrapper to match the time in force change

samuael committed:
fix time-in-force related test errors

samuael committed:
linter issue fix

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
time in force constants, functions and unit tests update

samuael committed:
shift tif policies to TimeInForce

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
linter fix and unit test config update

samuael committed:
Update time-in-force, related functions, and unit tests

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
fix linter issue and time-in-force processing

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
method and price type update

samuael committed:
Merge branch 'update_tif' into cryptoCom

samuael committed:
update wrapper with time-in-force

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
added a good till crossing tif value

samuael committed:
order type fix and fix related tim-in-force entries

samuael committed:
update time-in-force unmarshaling and unit test

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'update_tif' into cryptoCom

samuael committed:
fix time-in-force error in gateio

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
linter issue fix

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
update based on review comments

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
add unit test and fix missing issues

samuael committed:
minor fix and added benchmark unit test

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
change GTT to GTC for limit

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
fix linter issue

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
added time-in-force value to place order param

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'update_tif' into cryptoCom

samuael committed:
Fix linter issues

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
fix minor issues based on review comment and move tif code to separate files

samuael committed:
update on exchanges linked to time-in-force

samuael committed:
resolve missing review comments

samuael committed:
minor linter issues fix

samuael committed:
added time-in-force handler and update timeInForce parametered endpoint

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
minor fixes based on review

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
nits fix

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
update based on review

samuael committed:
linter fix

samuael committed:
rm getTimeInForce func and minor change to time-in-force

samuael committed:
minor change

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'update_tif' into cryptoCom

samuael committed:
update based on review comments

samuael committed:
wrappers and time-in-force calling approach

samuael committed:
minor change

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
update gateio string to timeInForce conversion and unit test

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
adding derivative endpoints and updating existing spot endpoints

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
updated order test unit tes functions

samuael committed:
add missing insurance endpoint and update websocket subscription

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
update account balance handling

samuael committed:
minor fixes on unit tests

samuael committed:
nits fix based on feedback

samuael committed:
update TestDeriveCancel unit test assert messages

samuael committed:
update TestDeriveCancel unit test assert messages

samuael committed:
adding websocket stream data processor for perpetual_swap instruments

samuael committed:
completed websocket push data handlers

samuael committed:
fix unit tests

samuael committed:
update timeInForceFromString method to return formatted error and update functions using it

samuael committed:
assets types update on wrapper and config files

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into update_tif

samuael committed:
reconfigure asset types and update wrapper functions and unit tests

samuael committed:
added a perpetualTP variable for testing perpetual swap instruments

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'update_tif' into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
update wrapper functions, helper functions, ratelimits, and unit tests

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
linter issues fix

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
resolve issues and overall update

samuael committed:
linter issues fix

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
rename orderbook Base to Book and Tranches to Level based on recent changes

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
fix based on orderbook changes

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
rename exchange receiver

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

dependabot[bot] committed:
build(deps): Bump github.com/stretchr/testify from 1.10.0 to 1.11.0 (#2013)

Bumps [github.com/stretchr/testify](https://github.com/stretchr/testify) from 1.10.0 to 1.11.0.
- [Release notes](https://github.com/stretchr/testify/releases)
- [Commits](https://github.com/stretchr/testify/compare/v1.10.0...v1.11.0)

---
updated-dependencies:
- dependency-name: github.com/stretchr/testify
  dependency-version: 1.11.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): Bump github.com/grpc-ecosystem/grpc-gateway/v2 (#2017)

Bumps [github.com/grpc-ecosystem/grpc-gateway/v2](https://github.com/grpc-ecosystem/grpc-gateway) from 2.27.1 to 2.27.2.
- [Release notes](https://github.com/grpc-ecosystem/grpc-gateway/releases)
- [Changelog](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/.goreleaser.yml)
- [Commits](https://github.com/grpc-ecosystem/grpc-gateway/compare/v2.27.1...v2.27.2)

---
updated-dependencies:
- dependency-name: github.com/grpc-ecosystem/grpc-gateway/v2
  dependency-version: 2.27.2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): Bump github.com/mattn/go-sqlite3 from 1.14.30 to 1.14.32 (#2015)

Bumps [github.com/mattn/go-sqlite3](https://github.com/mattn/go-sqlite3) from 1.14.30 to 1.14.32.
- [Release notes](https://github.com/mattn/go-sqlite3/releases)
- [Commits](https://github.com/mattn/go-sqlite3/compare/v1.14.30...v1.14.32)

---
updated-dependencies:
- dependency-name: github.com/mattn/go-sqlite3
  dependency-version: 1.14.32
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): Bump google.golang.org/protobuf from 1.36.7 to 1.36.8 (#2014)

Bumps google.golang.org/protobuf from 1.36.7 to 1.36.8.

---
updated-dependencies:
- dependency-name: google.golang.org/protobuf
  dependency-version: 1.36.8
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

samuael committed:
exchanges: Limit mock test JSON data size by truncating slices and maps (#1968)

* set limiter to first level mock data list and updated unit tests

* address nested slices length limit

* minor fix recording file and update unit tests

* minor updates on unit tests

* re-record mock files and minor fix on the unit tests ti adapt the mock data change

* improve http recording limit value and fix issues with mock data in binance

* added MockDataSliceLimit in request items and resolve minor unit test issues

* resolve missed conflict

* rename mock variables, resolve unit test issues, and other updates

* minor fix to CheckJSON and update unit tests

* minor unit test fix

* further optimization on mock CheckJSON method, unit tests, and re-record poloniex

* common and recording unit tests fix

* minor linter issues fix

* unit tests format fix

* fix miscellaneous error

* unit tests fix and minor docs update

* re-record and reduce mock file size

* indentation fix

* minor assertion test fix

* reverted log.Printf line in live testing

* rename variables

* update NewVCRServer unit test

* replace string comparison with *net.OpError check

* restructur net error test

* exchanges/mock: Remove redundant error assertion message in TestNewVCRServer

---------

Co-authored-by: Adrian Gallagher <adrian.gallagher@thrasher.io>

gloriousCode committed:
exchange/order/limits: Migrate to new package and integrate with exchanges (#1860)

* move limits, transition to key gen

* rollout NewExchangePairAssetKey everywhere

* test improvements

* self-review fixes

* ok, lets go

* fix merge issue

* slower value func,assertify,drop IsValidPairString

* remove binance reference for backtesting test

* Redundant nil checks removed due to redundancy

* Update order_test.go

* Move limits back into /exchanges/

* puts limits in a different box again

* SHAZBERT SPECIAL SUGGESTIONS

* Update gateio_wrapper.go

* fixes all build issues

* Many niteroos!

* something has gone awry

* bugfix

* gk's everywhere nits

* lint

* extra lint

* re-remove IsValidPairString

* lint fix

* standardise test

* revert some bads

* dupe rm

* another revert 360 mcgee

* un-in-revertify

* Update exchange/order/limits/levels_test.go

Co-authored-by: Adrian Gallagher <adrian.gallagher@thrasher.io>

* fix

* Update exchanges/binance/binance_test.go

HERE'S HOPING GITHUB FORMATS THIS CORRECTLY!

Co-authored-by: Gareth Kirwan <gbjkirwan@gmail.com>

* update text

* rn func, same line err gk4202000

---------

Co-authored-by: Adrian Gallagher <adrian.gallagher@thrasher.io>
Co-authored-by: Gareth Kirwan <gbjkirwan@gmail.com>

thrasher- committed:
linters: Add modernise tool check and fix issues (#2012)

* linters: Add modernise tool check and fix issues

* engine: Simplify exch.SetDefaults call and remove localWG

* CI: Revert config versions lint workflow

thrasher- committed:
devcontainer: Update Dockerfile and settings for VSCode (#2020)

* devcontainer: Update Dockerfile and configuration for GoCryptoTrader setup

* devcontainer: Update comment for clarity on format and config details

* devcontainer: Refactor post-create command to use dedicated script for tool installation

* devcontainer: Update comment for Go modernise tool for clarity

MathieuCesbron committed:
web: Fix typo in settings component (#2022)

Co-authored-by: mathieu cesbron <mathieucesbron@mathieus-MacBook-Pro.local>

samuael committed:
added missing endpoints, and fix linter and other issues

samuael committed:
updating trading endpoints based on recent changes

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
self review, minor fixes, and added announcement endpoint

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
adding fiat wallet api endpoints

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
completed staking endpoints

samuael committed:
added a single comment

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
fix errors caused after rebase

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
code optimization and wrapper update

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
minor change

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
refactor websocket handlers and wrappers

samuael committed:
linter issues fix and cleanup

samuael committed:
linter, codespell, and miscellaneous fix

samuael committed:
linter, codespell, and miscellaneous fix

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Fix config

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
misc minor fix

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
Merge branch 'master' of http://github.com/thrasher-corp/gocryptotrader into cryptoCom

samuael committed:
update the types and added test coverage

samuael committed:
fix misc error

samuael committed:
rename files

samuael committed:
config linter issue fix
